Add how pool works docs (#60)
radicle-dev/radicle-contracts: docs/how_pool_works.md

This document is an introduction to understanding how the funding pool contract works.
It doesn't describe all the implementation details or the API,
it focuses on the basic principles of the core mechanics.

# Overview

The funding pool is a smart contract, which creates real-time streams of donations.

> One can start, alter or end the process of sending his funds at any time with immediate effect.

Reviewer: "of sending their funds"

Author: Fixed

The flow of funds is automatically maintained and steady over time.

There are 3 roles present in the contract.
Any Ethereum address can be used in any number of these roles at the same time.

- The sender - has assets and chooses who does he want to send them to, how much and how fast
- The receiver - receives funds from senders

> - The proxy - receives funds from senders, but immediately passes them to receivers of his choice

Reviewer: "of their choice"

Author: Fixed

> - The proxy - receives funds from senders, but immediately passes them to receivers of their choice

Author: It means the speed of moving funds from the sender to the receiver

> - The proxy - receives funds from senders, but immediately passes them to receivers of their choice

Reviewer 2: exactly, I don't get what you mean by "speed"? As in, from the app perspective, I set how much I want to disburse "per month" across the receivers I have defined, not how fast the disbursement takes place. Could you elucidate me here? :)

> - The proxy - receives funds from senders, but immediately passes them to receivers of their choice

Author: Do you have an idea of what could be a substitute? `and how fast` => `and with what rate`? But then again what's a rate? The concept of speed seems understandable and precise enough for the fourth paragraph of the whole text, the details are laid out later.

> - The proxy - receives funds from senders, but immediately passes them to receivers of their choice

Reviewer 2: Is the "speed" here referring to how fast the pool balance will be emptied out? I think so, just want to clarify.
If so:

> Do you have an idea of what could be a substitute? `and how fast` => `and with what rate`?

I'd say `and at what rate`. What do you think?

> - The proxy - receives funds from senders, but immediately passes them to receivers of their choice

Author: Ok, fixed

## The cycles

The whole blockchain history is divided into cycles of equal length.
Every cycle contains the same, constant number of blocks.
In the examples below, we assume that the cycle length is 5.

![](how_pool_works_1.png)

# The receiver

The receiver requires no preparation, any address can receive donations at any time.

> The only function of this role is collection of funds sent by others.

Author: How is this version better?

> The only function of this role is collection of funds sent by others.

Reviewer 2: "The receiver requires no preparation,: is not clear. Preparation for what, by whom? What I suggest tries to make that clear. Moreover, it seems that the second bit of your first sentence and the second sentence merged and simplified communicate more concisely an clearly.

> The only function of this role is collection of funds sent by others.

Author: Ok, but I'll leave the second sentence, it's important to mention the collection step, the funds don't appear in the wallet automatically. Is it ok?

> The only function of this role is collection of funds sent by others.

Reviewer 2: Sounds good, Igor :)

RE the second sentence, there is a typo/missing word:

``` diff
The only function of this role is **the** collection of funds sent by others.
```

> The only function of this role is collection of funds sent by others.

Author: Fixed

## The deltas

On every block, a sender can send funds to a receiver.
The receiver can collect funds sent on a given block only when the cycle containing it is over.

![](how_pool_works_2.png)

Here we have a history of a receiver, who has 2 senders.
Each of them was sending different amounts over different periods of time.

> On the end of each cycle, the collectable amount was increased by the total sent amount.

Author: How is this version better?

> On the end of each cycle, the collectable amount was increased by the total sent amount.

Reviewer 2: The original text as misplaces commas and the writing style sound is as if you are actively presenting this, while that is not how it's read. Pus, a receiver doesn't "have" senders, they _receive_ funds (from senders).

> Each receiver needs to know, how much was sent to him.

Reviewer: "was sent to them"

Author: Fixed

To accomplish that the history of received funds needs to be stored in some form.
The receiver doesn't care, who exactly is sending on each block and in what configuration.
The only thing that matters are the collectable per-cycle values below the timeline.

> This reduces the amount of stored data significantly, but this can be optimized further.

Author: It's in the next paragraph

![](how_pool_works_3.png)

In this example, we started with having the raw collectable value of 23 for every block until

> the end of the blockchain.

Reviewer: what do you mean here "until the end of the blockchain?"

Author: I mean for infinite number of blocks higher than this one, like in "until the end of time". The blockchain obviously has no end :D

Reviewer 3: "End of time" might be clearer because to me, the end of the chain is the head/tip :stuck_out_tongue:

Author: Fixed

> the end of the blockchain.

Author: It's a description of data storage

We then reduced that to storing values added to the collectable amount on each cycle.

> Now we need to describe only cycles when receiving anything.

Author: No, storing where changes take place is the next step. Have you seen the picture?

> Now we need to describe only cycles when receiving anything.

Reviewer 2: I have, but can't understand what the goal fo this sentence is then.

> Now we need to describe only cycles when receiving anything.

Author: We're going from collectable, through added to a delta shaving off data on each step. Skipping the middle step would be confusing

> The senders usually are sending constant per-cycle amounts over long periods of time, so

Author: They can send over short periods too

> The senders usually are sending constant per-cycle amounts over long periods of time, so

Reviewer 2: I see. Isn't that describing things from a user point of view, which is sort of irrelevant here? I.e, we don't actually know for how long users will want to run their outgoing support, and as for the contracts, that information is not really important. What do you think?

> The senders usually are sending constant per-cycle amounts over long periods of time, so

Author: It's a rationalization of a storage design choice. This description is for the people reading the contract code, so I'd say, that it's important to have a proper mental model

the added values tend to create long series of constant numbers, in this case, 5s.

> We exploit that and turn them into deltas relative to the previous cycles.
> Now we need deltas only for cycles where the sending rate changes, it's very cheap.
> This is what the contract actually stores: a mapping from cycle numbers to deltas.

Reviewer 2: ```suggestion
We exploit that by storing them in the form of deltas relative to the previous cycles. 
That reduces storage costs since we only need to store the deltas for cycles where the sending rate changes.

This is, in fact, what the contract stores: a mapping from cycle numbers to deltas.
```

> a mapping from cycle numbers to deltas.

On this, isn't this map contextual? I.e, if we'd _only_ store that association, we couldn't figure the _from_ and _to_ of the stream. Could you expand on that or refresh my memory? :smile:

Author: What do you mean by contextual? The point is to NOT store the "from"

Reviewer 2: I mean that if there is no context and all we have is a cycle number and deltas, we'd have not enough data to actually perform a transfer.

Author: That's the point, cycle numbers and deltas ARE enough to describe a transfer from the sender's perspective

## Collecting

> When the receiver wants to collect the funds, he can iterate over the finished cycles,

Reviewer: "they can iterate over"

Author: Fixed

> reconstruct the added values from deltas and then calculate the final currently collectable value.

Reviewer: maybe better "the collectable amount at this point in time"

Author: Fixed

For any cycle before the first funding begins the added value is assumed to be 0.

# The sender

> The sender has a balance and a set of parameters, which describe the sending process.

Reviewer 2: ```suggestion
The sender has an available balance, a per-block rate, and a set of receivers. From the available balance,
the per-block rate is disbursed across all the receivers on every block.
```

?

The balance can be increased by topping up, which requires sending the assets to the contract.

> The currently unspent balance can be withdrawn and transferred out of the contract at any time.

Reviewer 2: ```suggestion
The currently remaining balance can be withdrawn and transferred out of the contract at any time.
```

What is the difference between "withdrawn" and "transferred out"?

Author: WIthdrawal is an action, the transfer is its underlying outcome, which results in moving funds out of the contract

Reviewer 2: Don't quite get the point. It seems that this says it succinctly:

```suggestion
The currently remaining balance can be withdrawn from the contract at any time.
```

The definition of `withdraw` encapsulates it all it seems to me.

> The overall rate of funding of the sender is described with the amount per block.
> It limits the value, which can be taken from the balance on each block and sent to the receivers.
> This value stays constant over time unless explicitly updated.

Reviewer 2: ```suggestion
The amount per block value defines the rate at which the funds are disbursed from the remaining balance 
to the receivers on each block. This value stays constant over time unless explicitly updated.
```

Author: This version is less precise

> The sender maintains a list of receivers too, each of them with a weight.
> The weights regulate how the per-block amount is split between the receivers.
> For example, a receiver with weight 2 is going to get a share twice as big
> as a receiver with weight 1, but only half as big as another receiver with weight 4.

Reviewer 2: ```suggestion
Finally, the sender maintains a list of receivers. 
Each receiver has an associated `weight`, which regulates how the amount per-block is 
split between the receivers.

For example, a receiver with weight 2 is going to get a share twice as big
as a receiver with weight 1, but only half as big as another receiver with weight 4.
```

Author: How is this better?

> ## Starting sending

Reviewer 2: I read [Starting Strength](https://startingstrength.com/) :laughing:

> When the sender has a sufficient balance, a sending rate and a list of receivers, sending starts.

Reviewer 2: ```suggestion
Given that a sender has sufficient balance, a sending rate per-block, and a list of receivers, sending starts automatically.
```

Author: How is that version better?

First, the funding period is calculated.
Its start is the current block and its end is the block on which the balance will run out.
Next, for each receiver the weighted share of the per-block rate is calculated.

> The receiver's deltas are updated to reflect that during the whole sending period on every block
> it's going to receive the calculated amount.

Reviewer 2: ```suggestion
The receiver's deltas are updated to the calculated amount that will be disbursed every block throughout the whole sending period.
```

Author: That's not true!

Let's take a look at an example of an application of a delta.
The sender will be sending 1 per block or 5 per cycle.

![](how_pool_works_4.png)

The deltas are applied relative to the existing values.

> It doesn't matter if anybody else is funding the receiver, it won't affect this sender.

Reviewer 2: ```suggestion
Note that this is a  timeline relative to a specific sender, therefore not considering possible other sends supporting this receiver.
```

Author: It's a timeline of a receiver, not a sender, it considers all the senders

Reviewer 2: sorry, my bad. updated.

Author: But there's still only one timeline per receiver, the only relative thing are the changes, but that's already said in the previous sentence

Another important point is that the delta changes are usually split between two cycles.
This reflects that the first cycle is only partially affected by the change in funding.

> Only the second one is fully affected and it must apply the rest of the delta.

Reviewer 2: Very clear :ok_hand:

In this case, the total change of the per-cycle delta is +5 to start sending.
The current cycle isn't fully affected though, only 2 out of 5 blocks are sending.

> It's effectively going to transfer only the amount of 2, which is reflected in the +2 delta change.
> On the other hand, the next cycle and the ones after it are going to transfer the full 5.
> This is expressed with the +3 delta change, which turns 2 per cycle into the full 5 per cycle.

Reviewer 2: ```suggestion
It's effectively only going to transfer only the amount of 2 in this current cycle, which is reflected in the +2 delta change.
From the next block onwards, the amount of 5 will be disbursed. This is expressed with the +3 delta change, 
which turns the previous 2 per cycle into the full 5 per cycle.
```

Author: How is this version better?

> ## Stopping sending

Reviewer 2: On this section, it would be more digestible to first present what we consider "stopping sending", and then explain the mechanisms through which we represent that. Maybe starting with what is now the last paragraph would already help a lot.

Author: Which paragraph do you mean?

Reviewer 2: I mean the last paragraph of this section.

Author: Fixed

When funding stops, the deltas need to be reverted.
To do that basically the same process is applied, just with negative deltas.

In this case, the reverting is split into 2 cycles too, one with -4 and the other with -1.

Let's assume that a few blocks have passed, but the sender wants to stop sending.
This can happen because the sender doesn't want to fund the receiver anymore
or because they want to change some of its configuration.
In the latter case sending is stopped only to be immediately resumed, but with different parameters.
Either way, the effects of the sender on the receiver's deltas need to be reverted
from the current block to the end of the existing funding period.

![](how_pool_works_5.png)

The old funding end deltas are reverted because they don't reflect the real funding end anymore.
On the other hand, a new end is applied to the current block,
just as if it was always supposed to be the end of the funding period.
Now the receiver's future isn't affected by the sender anymore.
The past stays untouched because the already sent funds are out of the sender's control.

# The proxy

> The proxy is configured only with a list of weighted receivers.

Reviewer 2: ```suggestion
The proxy is configured only with a list of receivers with an associated weight.
```

We have already mentioned above that a receiver has a weight. Do we need to repeat that here? Maybe yes, just wan to double-check the goal here.

Author: We described above the sender's receivers, the proxy could have no weights, so it's important to mention

> It's not possible to send funds via a proxy which isn't configured.

Reviewer 2: What is a non-configured proxy, or what would that look like?

Author: Fixed

> The sum of the receivers' weights must always be a constant value,

Reviewer 2: > weights must always be a constant value,

under which context? Not sure I understand? If I start with `[(a, 1), (b, 1)]` and then change it to `[(a, 1), (b, 2)]`, the sum of the weights is no longer the same as it was before. That's how I mean my question :)

Author: You're right, if you try to make such change, you break the rule about the sum of weights, which is described here. There's no context, it's a global rule. What's your question?

which is defined in the contract and it's the same for all the proxies.

> Just like a receiver, the proxy has a mapping between the cycles and the deltas of received amounts.

Reviewer 2: ```suggestion
Just like a receiver, the proxy has a mapping between the cycles and the deltas of the amounts to disburse.
```

Author: The receiver doesn't disburse amounts, so that change would break this analogy

Reviewer 2: I don't think this sentence implies that the receiver is the one who disburses, but that the proxy has a mapping similar to the receivers.

Author: After thorough reading you're right, it means amounts to disburse INTO a receiver/proxy. Is disburse a better word here than receive?

Reviewer 2: Cool :)

> Is disburse a better word here than receive?

Since the flow here is proxy -> receivers I'd say "disburse", yeah.

Author: I was viewing it as sener -> receiver, it's more consistent with the overall narration

Reviewer 2: Yes you are right :) "receive" then :100:

## Sending via a proxy

When a sender starts sending funds to a proxy, it does so in two steps.
First it applies changes to the proxy's deltas similarly to how it would do with a receiver.
Next, it iterates over all the proxy's receivers and applies changes to
their deltas as if they were directly funded by the sender.
The funding rate applied to the receivers is split according to their weights in the proxy.

> For the example sake let's assume that the delta's proxy weights sum must be equal 5.

Reviewer 2: ```suggestion
For the example sake let's assume that the delta's proxy weights sum must be equal to 5.
```

The proxy has 2 receivers: A with weight 3 and B with weight 2.
The sender wants to start sending via the proxy 2 per block or 10 per cycle.
It's 2 per cycle per proxy weight.

![](how_pool_works_6.png)

The proxy's deltas store amount per 1 proxy weight, which is 2.
The receivers get their shares, A with weight 3 gets 6 and B with weight 2 gets 4 per cycle.

When a sender stops sending, the process is reversed like with regular receivers.
All the deltas are once again applied, but with negative values.

## Updating a proxy

When the list of proxy receivers is updated, all funding must be moved to a new set of receivers.
That's when the proxy's deltas come useful.

> For each cycle and each receiver, the proxy can tell, what total delta it has applied.

Reviewer 2: ```suggestion
For each cycle and each receiver, the proxy can tell the total delta it has applied.
```

It can then use this information to erase its influence from its receivers.

![](how_pool_works_7.png)

In this example, the receiver's weight is 3.
To erase the influence, the proxy's deltas are multiplied by
the receiver's weights and subtracted from the corresponding receiver's deltas.

After removing its influence from one set of receivers the proxy must reapply itself on a new set.
This is done in the same way as removal, but this time the deltas are added and not subtracted.

### The current cycle problem

Unlike the senders, the proxies store data with a per-cycle precision.
When changing the set of receivers, a delta describing the current cycle may need to be applied.

> When it happens, it's unclear, what part of the per-cycle delta should be moved,

Reviewer 2: ```suggestion
When it happens, it's unclear what part of the per-cycle delta should be moved,
```

because some funds were sent before the current block and some will be sent after it.

![](how_pool_works_8.png)

The solution is to ignore the problem and move the whole current cycle delta.

> Some funds already sent in the current cycle may disappear from one receiver and appear in another,
> but it's not dangerous, the receivers have no access to funds coming from an unfinished cycle.

Reviewer 2: ```suggestion
Some funds already sent in the current cycle may disappear from one receiver and appear in another.
Such behaviour, however, is not of significant importance since the receivers have no access to funds 
coming from an unfinished cycle.
```

This document is an introduction to understanding how the funding pool contract works.
It doesn't describe all the implementation details or the API,
it focuses on the basic principles of the core mechanics.

# Overview

The funding pool is a smart contract which creates real-time streams of donations.
One can start, alter or end the process of sending their funds at any time with immediate effect.
The flow of funds is automatically maintained and steady over time.

There are 3 roles present in the contract.
Any Ethereum address can simultaneously take on any of these roles.

- **The sender**: has assets and chooses who do they want to send them to,
how much, and at what rate
- **The receiver:** receives funds from senders
- **The proxy**: receives funds from senders,
but immediately passes them to receivers of their choice

## The cycles

> The whole blockchain history is divided into cycles of equal length.

Author: Fixed

> Every cycle contains the same, constant number of blocks.

Author: Fixed

In the examples below, we assume that the cycle length is 5.

![](how_pool_works_1.png)

# The receiver

> There no setup steps for one to become a receiver.

Reviewer 3: ```suggestion
There are no setup steps for one to become a receiver.
```

Any address can receive donations at any time, from any sender.

> The only function of this role is collection of funds sent by others.

Reviewer 2: ```suggestion
The only function of this role is the collection of funds sent by others.
```

Author: I did pass it through Grammarly before submitting the PR and fixed most of the issues. I remember this exact issue and I remember explicitly NOT fixing it. Grammarly treated `collection` as a regular noun, but here it's a noun derived from a verb, which is too abstract to use `the` here.

Reviewer 2: Not sure if it makes a difference but I use a paid version and there are currently 21 correctness alerts left, excluding this one.

Author: I've used the free version, which mostly checks the grammar. The premium features are mostly about style, which may or may not be even relevant here. I got 69 issues in total, but with access to details about only 11.

Reviewer 4: Without “the” the sentence sound off to me.

## The deltas

On every block, a sender can send funds to a receiver.

> The receiver can collect funds sent on a given block only when the cycle containing it is over.

Reviewer 4: We should elaborate on this
* The “collect” transaction triggered by a receiver moves all collectable funds to the recipients account
* We need to explain how the collectable funds are calculated
* We need to explain that collecting funds reduces the collectable funds.

Author: Fixed

The only thing that matters are the collectable per-cycle values below the timeline.
This reduces the amount of stored data significantly, but this can be optimized further.

![](how_pool_works_3.png)

In this example, we started with having the raw collectable value of 23 for every block until
the end of time.
We then reduced that to storing values added to the collectable amount on each cycle.
Now we need to describe only cycles when receiving anything.
The senders usually are sending constant per-cycle amounts over long periods of time, so
the added values tend to create long series of constant numbers, in this case, 5s.
We exploit that and turn them into deltas relative to the previous cycles.
Now we need deltas only for cycles where the sending rate changes, it's very cheap.
This is what the contract actually stores: a mapping from cycle numbers to deltas.

## Collecting

When the receiver wants to collect the funds, they can iterate over the finished cycles, reconstruct
the added values from deltas and then calculate the collectable amount at this point in time.
For any cycle before the first funding begins the added value is assumed to be 0.

> # The sender

Author: Fixed

> The sender has an available balance, a per-block rate, and a set of receivers.

Reviewer 4: * We need to find the right terminology for the balance. Is “available” needed? What is the “unavailable” part of the balance. How does it relate to the “currently remaining balance” mentioned below?
* Instead of “per-block rate” I’d suggest the term “funding rate” and then explain that it is expressed in tokens (or funds or something else) per block.

Author: Both fixed

> From the available balance, the per-block rate is disbursed across all the receivers on every block.

Reviewer 4: We should make the relation ship between the balance an  rate clearer: The balance decreases every block by the funding rate. The funds are credited to the receivers. I’d also avoid the term “disbursed ... on every block” because it sounds like this is an explicit action and one might have the impression that this changes something in the chain state while it is only a bookkeeping thing.

Author: Fixed

> The balance can be increased by topping up, which requires sending the assets to the contract.

Reviewer 4: We should clearly point out the actions that a sender can take and mention that explicitly instead of using passive voice.

Author: Fixed

The currently remaining balance can be withdrawn from the contract at any time.

The overall rate of funding of the sender is described with the amount per block.
It limits the value, which can be taken from the balance on each block and sent to the receivers.
This value stays constant over time unless explicitly updated.

> The sender maintains a list of receivers too, each of them with a weight.

Reviewer 4: Why “too”? Does this reference another thing that maintains a list of receivers? We’re already mentioning receivers at the beginning of the section so I don’t think we need the “too”.

Author: Fixed

It doesn't describe all the implementation details or the API,
it focuses on the basic principles of the core mechanics.

# Overview

The funding pool is a smart contract which creates real-time streams of donations.
One can start, alter or end the process of sending their funds at any time with immediate effect.
The flow of funds is automatically maintained and steady over time.

There are 3 roles present in the contract.
Any Ethereum address can simultaneously take on any of these roles.

- **The sender**: has assets and chooses who do they want to send them to,
how much, and at what rate
- **The receiver:** receives funds from senders
- **The proxy**: receives funds from senders,
but immediately passes them to receivers of their choice

## The cycles

The whole blockchain history is divided into cycles of
equal length so that every block belongs to a cycle.
Cycles are numbered starting with zero.
In the examples below, we assume that the cycle length is 5.

![](how_pool_works_1.png)

# The sender

The sender has a balance, a funding rate, and a set of receivers.

The balance is automatically reduced by the funding rate on every block

> and the same amount is credited to the sender's receivers.

Author: Fixed

This process doesn't actually require updates on every block,
its effects are calculated on the fly whenever they are needed.
Thus the contract state is updated only when the funding parameters are altered by the users.

The sender balance is manually increased by topping up,
which requires sending the assets from the user wallet to the contract.
The opposite operation is withdrawal,
which results in sending the assets from the contract back to the user wallet.

This funding rate stays constant over time unless explicitly updated.

The sender maintains a list of receivers, each of them with a weight.
The weights regulate how the funded amount is split between the receivers.
For example, a receiver with weight 2 is going to get a share twice as big
as a receiver with weight 1, but only half as big as another receiver with weight 4.

> ## The deltas

Author: Agreed, it's cluttered and murky. Applied your proposition with some changes to fit the surrounding text.

> On every block, a sender sends funds to a receiver.

Author: Applied

Each receiver needs to know, how much was sent to them.
To accomplish that the history of received funds needs to be stored in some form.
The receiver doesn't care, who exactly is sending on each block and in what configuration.
The only thing that matters are the collectable per-cycle values below the timeline.
This reduces the amount of data stored per receiver significantly, but this is further optimized.

![](how_pool_works_3.png)

In this example, we start with having the raw **collectable** value of 23 for every block until
the end of time.
Next, we reduce that to storing values **added** to the collectable amount on each cycle.
Now we need to describe only cycles when receiving anything.
The senders usually are sending constant per-cycle amounts over long periods of time, so
the added values tend to create long series of constant numbers, in this case, 5s.
We exploit that and finally turn them into **deltas** relative to the previous cycles.
Now we need to store data only for cycles where the sending rate changes, it's very cheap.
This is what the contract actually stores: a mapping from cycle numbers to deltas.

## Starting sending

> When the sender has a sufficient balance, a sending rate and a list of receivers,

Author: All fixed

> the flow of assets automatically starts.

Author: Fixed

from the current block to the end of the existing funding period.

![](how_pool_works_5.png)

The old funding end deltas are reverted because they don't reflect the real funding end anymore.
On the other hand, a new end is applied to the current block,
just as if it was always supposed to be the end of the funding period.
Now the receiver's future isn't affected by the sender anymore.
The past stays untouched because the already sent funds are out of the sender's control.

# The receiver

There are no setup steps for one to become a receiver.
Any address can receive donations at any time, from any sender.
The only function of this role is the collection of funds sent by others.

## Collecting

The receiver can at any time collect the funds sent to it.
The contract calculates the total amount and then transfers it out to the receiver's wallet.
The collected amount is always everything available at a given moment, there's no way to limit it.

As shown in the previous sections, the collectable amount is described with deltas, one per cycle.
The receiver stores the number of the first cycle, for which the funds haven't been collected yet.

> This prevents collecting the same funds multiple times, every cycle can be ever processed only once.

Author: Good point, fixed

The receiver also stores the amount, which was collected for the last collected cycle.
This value is set to 0 if this is the first collection of the receiver.
It's the initial value to which the deltas are added.

To calculate, how much the receiver can collect,
the contract iterates over all the completed cycles since the first uncollected one.
For each of them, it adds the corresponding delta to the value collected from the previous cycle.
This reconstructs the amount sent from all the senders during each cycle.
These amounts are then added up and the result is the total collected amount.
Finally, the next uncollected cycle number and the last cycle amount are updated.

![](how_pool_works_6.png)

In this example funds received from 4 cycles are being collected.
The yellow fields are the stored state before the collection, green after it.
The blue field is the collected value, which is going to be transferred to the sender's wallet.

> # The proxy

Author: Added

First it applies changes to the proxy's deltas similarly to how it would do with a receiver.
Next, it iterates over all the proxy's receivers and applies changes to
their deltas as if they were directly funded by the sender.
The funding rate applied to the receivers is split according to their weights in the proxy.

For the example sake let's assume that the delta's proxy weights sum must be equal to 5.
The proxy has 2 receivers: A with weight 3 and B with weight 2.
The sender wants to start sending via the proxy 2 per block or 10 per cycle.
It's 2 per cycle per proxy weight.

![](how_pool_works_7.png)

The proxy's deltas store amount per 1 proxy weight, which is 2.
The receivers get their shares, A with weight 3 gets 6 and B with weight 2 gets 4 per cycle.

When a sender stops sending, the process is reversed like with regular receivers.
All the deltas are once again applied, but with negative values.

## Updating a proxy

When the list of proxy receivers is updated, all funding must be moved to a new set of receivers.
That's when the proxy's deltas come useful.
For each cycle and each receiver, the proxy can tell the total delta it has applied.

> It can then use this information to erase its influence from its receivers.

Author: Fair, fixed
